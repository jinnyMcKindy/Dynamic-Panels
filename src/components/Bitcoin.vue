<template>
  <v-container>
    <v-row>
      <v-col cols="12">
        <router-link class="nav-link nav-link--light" :to="{ name: 'modals' }">
          Go to Modals
        </router-link>
      </v-col>
    </v-row>
    <v-row>
      <v-col cols="3">
        <v-btn small color="primary" @click="start">Запуск</v-btn>
      </v-col>
      <v-col cols="3">
        <v-btn small color="error" @click="stop">Остановка</v-btn>
      </v-col>
      <v-col cols="3">
        <v-btn small color="warning" @click="clear">Сброс</v-btn>
      </v-col>
    </v-row>
    <v-row>
      <v-col cols="12">
        <h3>Сумма: {{ total }} BTC</h3>
      </v-col>
    </v-row>
    <v-row>
      <v-col cols="12">
        <v-simple-table>
          <template v-slot:default>
            <thead>
              <tr>
                <th class="text-left">From</th>
                <th class="text-left">To</th>
                <th class="text-left">Sum</th>
              </tr>
            </thead>
            <tbody>
              <tr v-for="(item, index) in transactions" :key="index">
                <td>
                  <div v-for="(from, key) in item.from" :key="key">
                    {{ from }}
                  </div>
                </td>
                <td>
                  <div v-for="(to, key) in item.to" :key="key">
                    {{ to }}
                  </div>
                </td>
                <td>{{ item.sum }} BTC</td>
              </tr>
            </tbody>
          </template>
        </v-simple-table>
        <div>{{ error }}</div>
      </v-col>
    </v-row>
  </v-container>
</template>

<script>
export default {
  name: "Bitcoin",
  data: function() {
    return {
      error: "",
      transactions: [],
      socket: new WebSocket("wss://ws.blockchain.info/inv")
    };
  },
  computed: {
    total: function() {
      return this.transactions.reduce(
        (acc, val) => (acc += parseFloat(val.sum)),
        0
      );
    }
  },
  mounted() {
    this.socket.onmessage = event => {
      const data = JSON.parse(event.data)["x"];
      const from = [];
      const to = [];
      let satoshi = 0;
      data.inputs.forEach(prev => {
        from.push(prev.prev_out.addr);
      });
      data.out.forEach(out => {
        to.push(out.addr);
        satoshi += parseFloat(out.value);
      });
      const sum = parseFloat(satoshi) / 100000000;
      this.transactions.push({ from, to, sum });
    };

    this.socket.onclose = function(event) {
      if (event.wasClean) {
        this.error = `[close] Соединение закрыто чисто, код=${event.code} причина=${event.reason}`;
      } else {
        this.error = "[close] Соединение прервано";
      }
    };

    this.socket.onerror = function(error) {
      this.error = `[error] ${error.message}`;
    };
  },
  methods: {
    start() {
      this.socket.send(JSON.stringify({ op: "unconfirmed_sub" }));
    },
    stop() {
      this.socket.send(JSON.stringify({ op: "unconfirmed_unsub" }));
    },
    clear() {
      this.transactions = [];
    }
  }
};
</script>
