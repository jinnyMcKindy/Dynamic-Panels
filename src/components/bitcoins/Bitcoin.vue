<template>
  <v-container>
    <v-row>
      <v-col cols="12">
        <router-link class="nav-link nav-link--light" :to="{ name: 'modals' }">
          Go to Modals
        </router-link>
      </v-col>
    </v-row>
    <v-row>
      <v-col cols="3">
        <Button color="primary" :disabled="btn.start" @click="start"
          >Запуск</Button
        >
      </v-col>
      <v-col cols="3">
        <Button color="error" :disabled="!btn.start" @click="stop"
          >Остановка</Button
        >
      </v-col>
      <v-col cols="3">
        <Button color="warning" :disabled="!transactions.length" @click="clear"
          >Сброс</Button
        >
      </v-col>
    </v-row>
    <v-row>
      <v-col cols="12">
        <h3>Сумма: {{ total }} BTC</h3>
      </v-col>
    </v-row>
    <v-row>
      <v-col cols="12">
        <Table :transactions="transactions" :error="error" />
      </v-col>
    </v-row>
  </v-container>
</template>

<script>
import Button from "./Button";
import Table from "./Table";
export default {
  name: "Bitcoin",
  components: { Button, Table },
  data: function() {
    return {
      error: "",
      transactions: [],
      socket: new WebSocket("wss://ws.blockchain.info/inv"),
      btn: {
        start: false
      }
    };
  },
  computed: {
    total: function() {
      return this.transactions.reduce(
        (acc, val) => (acc += parseFloat(val.sum)),
        0
      );
    }
  },
  mounted() {
    this.socket.onmessage = event => {
      const data = JSON.parse(event.data)["x"];
      const from = [];
      const to = [];
      let satoshi = 0;
      data.inputs.forEach(prev => {
        from.push(prev.prev_out.addr);
      });
      data.out.forEach(out => {
        to.push(out.addr);
        satoshi += parseFloat(out.value);
      });
      const sum = parseFloat(satoshi) / 100000000;
      this.transactions.push({ from, to, sum });
    };

    this.socket.onclose = function(event) {
      if (event.wasClean) {
        this.error = `[close] Соединение закрыто чисто, код=${event.code} причина=${event.reason}`;
      } else {
        this.error = "[close] Соединение прервано";
      }
    };

    this.socket.onerror = function(error) {
      this.error = `[error] ${error.message}`;
    };
  },
  methods: {
    start() {
      this.socket.send(JSON.stringify({ op: "unconfirmed_sub" }));
      this.btn.start = true;
    },
    stop() {
      this.socket.send(JSON.stringify({ op: "unconfirmed_unsub" }));
      this.btn.start = false;
    },
    clear() {
      this.transactions = [];
    }
  }
};
</script>
