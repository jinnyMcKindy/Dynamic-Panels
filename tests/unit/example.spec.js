import { createLocalVue, mount } from "@vue/test-utils";
import Bitcoin from "@/components/bitcoins/Bitcoin.vue";
import Vuetify from 'vuetify'

describe("HelloWorld.vue", () => {
  const localVue = createLocalVue()
  let vuetify

  beforeEach(() => {
    vuetify = new Vuetify()
  })
  it("renders props.msg when passed", () => {
    const wrapper = mount(Bitcoin, {
      localVue,
      vuetify,
    })
    expect(wrapper.exists()).toBe(true)
  });
});
