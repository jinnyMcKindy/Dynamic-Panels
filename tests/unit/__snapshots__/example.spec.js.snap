// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HelloWorld.vue renders props.msg when passed 1`] = `
<v-container>
  <v-row>
    <v-col cols="12">
      <router-link to="[object Object]" class="nav-link nav-link--light">
        Go to Modals
      </router-link>
    </v-col>
  </v-row>
  <v-row>
    <v-col cols="3">
      <v-btn small="" color="primary">Запуск</v-btn>
    </v-col>
    <v-col cols="3">
      <v-btn small="" color="error" disabled="true">Остановка</v-btn>
    </v-col>
    <v-col cols="3">
      <v-btn small="" color="warning" disabled="true">Сброс</v-btn>
    </v-col>
  </v-row>
  <v-row>
    <v-col cols="12">
      <h3>Сумма: 0 BTC</h3>
    </v-col>
  </v-row>
  <v-row>
    <v-col cols="12">
      <div>
        <v-simple-table></v-simple-table>
        <div></div>
      </div>
    </v-col>
  </v-row>
</v-container>
`;
